### PR TITLE
add: ollama keep alive parameter added. issue #4024

### LINKED ISSUE #4024: support for ollama keep_alive parameter
### Self Checks

- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] Pleas do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing?

yes

### 2. Describe the feature you'd like to see

support the keep_alive feature of ollama, detailed in https://github.com/ollama/ollama/pull/2146

### 3. How will this feature improve your workflow or experience?

currently the model will unload after 5 min idle, the next reload will take more than 30min for a 70b model.

### 4. Additional context or comments

no

### 5. Can you help us with this feature?

- [ ] I am interested in contributing to this feature.

### PR DESCRIPTION
**Pull request by Yash-1511**
# Description

Adding Ollama Keep alive parameter to adjust to control the duration for how long a model is loaded and left in memory. 

Fixes #4024 

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# How Has This Been Tested?

I have tested via sending requests with different different times and watch the activity monitor to load and unload of the model. 
- [ ] TODO

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods
- [ ] `optional` I have made corresponding changes to the documentation 
- [ ] `optional` I have added tests that prove my fix is effective or that my feature works
- [ ] `optional` New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by QIN2DIM** [comment]
Are you using the latest version of ollama? How did you configure these parameters?
- `OLLAMA_KEEP_ALIVE`
- `OLLAMA_NUM_PARALLEL`
- `OLLAMA_MAX_LOADED_MODELS`

With these feature updates ollama can handle parallel requests and keep models in memory for any length of time. 

Previously, the default value of OLLAMA_KEEP_ALIVE was 5 minutes. In the Dify workflow, I often encountered the situation where ollama re-read the model, which took a lot of time. But after I reconfigured these parameters, all the problems were solved.

In other words, if you want the model to be memory-resident, perhaps `OLLAMA_KEEP_ALIVE=-1` would be more appropriate? It may be closer to best practice to modify the configuration directly in ollama.services.

**Comment by Yash-1511** [comment]
> @Yash-1511 Are you using the latest version of ollama. How did you configure these parameters:
> 
> * `OLLAMA_KEEP_ALIVE`
> * `OLLAMA_NUM_PARALLEL`
> * `OLLAMA_MAX_LOADED_MODELS`
> 
> With these feature updates ollama can handle parallel requests and keep models in memory for any length of time.
> 
> Previously, the default value of `OLLAMA_KEEP_ALIVE` was 5 minutes. In the Dify workflow, I often encountered the situation where ollama re-read the model, which took a lot of time. But after I reconfigured these parameters, all the problems were solved.
> 
> In other words, if you want the model to be memory-resident, perhaps `OLLAMA_KEEP_ALIVE=-1` would be more appropriate? It may be closer to best practice to modify the configuration directly in ollama.services.

Certainly! Here's a rephrased version of your message:

---

Hello @QIN2DIM,

Yes, I'm using the latest version of Ollama.

For the `OLLAMA_KEEP_ALIVE` parameter, it can be configured in several ways:

- As a duration string, like "10m" or "24h".
- As a number in seconds, such as 3600.
- As any negative number (e.g., -1 or "-1m"), which will keep the model loaded indefinitely.
- As '0', which will unload the model immediately after generating a response.

In Dify, I've added a new field for the `keep_alive` parameter, allowing you to set durations as described. For instance, you can use -1 to keep the model in memory permanently, 0 to unload it right after a response, or specify any duration in seconds or minutes.

Regarding the other two parameters you mentioned, `OLLAMA_NUM_PARALLEL` and `OLLAMA_MAX_LOADED_MODELS`, these are still experimental. (release 0.1.33):

- `OLLAMA_NUM_PARALLEL` allows handling multiple requests simultaneously for a single model.
- `OLLAMA_MAX_LOADED_MODELS` enables loading multiple models at once.

To activate these features, you'll need to set the environment variables when running the Ollama server. For more detailed information, you can refer to the [Ollama FAQ](https://github.com/ollama/ollama/blob/main/docs/faq.md#how-do-i-configure-ollama-server).

I hope this helps!

**Review by crazywoola — DISMISSED** [review]

**Comment by crazywoola** [comment]
@Yeuoly Please take a look at this.

File: api/core/model_runtime/model_providers/ollama/llm/llm.py [hunk 1]
```diff
@@ -544,56 +630,77 @@ def get_customizable_model_schema(self, model: str, credentials: dict) -> AIMode
...
                 ),
                 ParameterRule(
-                    name='format',
+                    name="keep_alive",
+                    label=I18nObject(en_US="Keep Alive"),
+                    type=ParameterType.STRING,
+                    help=I18nObject(
+                        en_US="Sets how long the model is kept in memory after generating a response. "
+                        "This can be a duration string (e.g., '10m' or '24h'), a number in seconds "
+                        "(e.g., 3600), a negative number to keep the model loaded indefinitely, or '0' "
```
Comment (Yeuoly) [inline api/core/model_runtime/model_providers/ollama/llm/llm.py]: Hi, I found that number without units could not be accepted, there is an error raised
image
And I've also checked the source code of Ollama, it uses json.Unmarshal to parse requests and `keep_alive` is defined as a time.Duration object, it requires at least one unit to be unmarshalled
image
Maybe we could optimize this description or just append `s` to `keep_alive` if there is not a unit?

**Review by Yeuoly — CHANGES_REQUESTED** [review]

File: api/core/model_runtime/model_providers/ollama/llm/llm.py (hunk 1 above)
Comment (Yash-1511) [reply api/core/model_runtime/model_providers/ollama/llm/llm.py]: I have updated the helper text.

**Review by Yeuoly — APPROVED** [review]
LGTM

**Comment by Yeuoly** [comment]
Thx :)
